src/charm.py

```diff
@@ -28,9 +30,104 @@ def __init__(self, *args: typing.Any):
+            self.on.nts_certificates_relation_broken, self._on_certificates_relation_broken
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_available, self._on_certificate_available
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_expiring, self._on_certificate_expiring
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_invalidated, self._on_certificate_invalidated
+        )
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_unit_status)
+
+    def _on_collect_unit_status(self, _: ops.CollectStatusEvent) -> None:
+        """Set unit status based on current charm status."""
+        if not self._get_time_sources():
+            self.unit.status = ops.BlockedStatus("no time source configured")
+            return
+        if not self._get_server_name() and self.model.get_relation("nts-certificates"):
+            self.unit.status = ops.BlockedStatus(
+                "server-name is required for nts-certificates integration"
+            )
+            return
+        if self.certificates.get_certificate_signing_requests(unfulfilled_only=True):
+            self.unit.status = ops.WaitingStatus("waiting for new certificate")
+            return
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_certificates_relation_created(self, _: ops.RelationCreatedEvent) -> None:
+        """Handle the certificates relation-creation event."""
+        self.unit.open_port("tcp", 4460)
+        self._do_renew_certificate()
+
+    def _on_certificates_relation_broken(self, _: ops.RelationBrokenEvent) -> None:
+        """Handle the certificates relation-broken event."""
+        self.unit.close_port("tcp", 4460)
+        self.tls_keychain.clear()
+        self._configure_chrony()
+
+    def _on_certificate_expiring(self, _: ops.EventBase) -> None:
+        """Handle the certificates expiring event."""
+        self._do_renew_certificate()
+
+    def _on_certificate_invalidated(self, _: ops.EventBase) -> None:
+        """Handle the certificates invalidated event."""
+        self._do_renew_certificate()
+
+    def _do_renew_certificate(self) -> None:
+        """Handle the event when certificate is unavailable."""
+        if not self._get_server_name():
+            return
+        self._renew_certificate()
+
+    def _on_certificate_available(self, event: tls_certificates.CertificateAvailableEvent) -> None:
+        """Handle the certificate-available event.
+
+        Args:
+            event: certificate-available event object.
+        """
+        self.tls_keychain.set_chain(event.chain_as_pem().strip())
+        self._configure_chrony()
+
+    def _renew_certificate(self) -> None:
+        """Renew the certificate."""
+        old_csr = self.tls_keychain.get_csr()
+        private_key = self.tls_keychain.get_private_key()
+        assert self._get_server_name()  # nosec
```

Comment:
Nitpick: Im wondering if this should be a defined error than an assert

Reply:
Yes, there are some discussions above https://github.com/canonical/chrony-operator/pull/4#discussion_r1673715461. I changed this to an `AssertError` based on @cbartz 's suggestion.